pkg/util/servicediscovery/ring.go

```diff
@@ -53,43 +59,65 @@ func (r *ringServiceDiscovery) running(ctx context.Context) error {
 	defer ringTicker.Stop()
 
 	// Notifies the initial state.
-	ringState, _ := r.ringClient.GetAllHealthy(r.ringOp) // nolint:errcheck
-	r.notifyChanges(ringState)
+	all, _ := r.ringClient.GetAllHealthy(activeRingOp) // nolint:errcheck
+	r.notifyChanges(all)
 
 	for {
 		select {
 		case <-ringTicker.C:
-			ringState, _ := r.ringClient.GetAllHealthy(r.ringOp) // nolint:errcheck
-			r.notifyChanges(ringState)
+			all, _ := r.ringClient.GetAllHealthy(activeRingOp) // nolint:errcheck
+			r.notifyChanges(all)
 		case <-ctx.Done():
 			return nil
 		case err := <-r.subservicesWatcher.Chan():
-			return errors.Wrap(err, "a subservice of ring-based service discovery has failed")
+			return errors.Wrap(err, "a subservice of query-schedulers ring-based service discovery has failed")
 		}
 	}
 }
 
-// notifyChanges is not concurrency safe.
-func (r *ringServiceDiscovery) notifyChanges(discovered ring.ReplicationSet) {
-	// Build a map with the discovered addresses.
-	discoveredAddresses := make(map[string]struct{}, len(discovered.Instances))
-	for _, instance := range discovered.Instances {
-		discoveredAddresses[instance.GetAddr()] = struct{}{}
+// notifyChanges is not concurrency safe. The input all and inUse ring.ReplicationSet may be the same object.
+func (r *ringServiceDiscovery) notifyChanges(all ring.ReplicationSet) {
+	// Build a map with the discovered instances.
```

Comment:
nit: All instances are "discovered" (in the ring). Perhaps we can just call this "instancesByAddress".